metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,458 @@
+            connectionsToAcquire.offer(updatedRequest)
+          )
+        case Right(connection) =>
+          connectionsRegister
+            .registerConnection(connection.encryptedConnection)
+            .flatMap(_ =>
+              connectionsQueue.offer(connection.encryptedConnection)
+            )
+      }
+      .completedF
+  }
+
+  /** Reads incoming connections in linear fashion and check if they are on cluster allowed list.
+    */
+  private def handleServerConnections[F[_]: Concurrent: TaskLift, K, M: Codec](
+      pg: EncryptedConnectionProvider[F, K, M],
+      connectionsQueue: ConcurrentQueue[F, EncryptedConnection[F, K, M]],
+      connectionsRegister: ConnectionsRegister[F, K, M],
+      clusterConfig: ClusterConfig[K]
+  ): F[Unit] = {
+    Iterant
+      .repeatEvalF(pg.incomingConnection)
+      .takeWhile(_.isDefined)
+      .map(_.get)
+      .collect { case Right(value) =>
+        value
+      }
+      .mapEval { encryptedConnection =>
+        if (
+          clusterConfig.isAllowedIncomingConnection(
+            encryptedConnection.remotePeerInfo._1
+          )
+        ) {
+          connectionsRegister
+            .isConnectionAlreadyRegistered(encryptedConnection)
+            .flatMap { alreadyRegistered =>
+              // this case can arise when incoming remote peer fails without giving any notice
+              // after restart such peer will try to connect to our local node, even though local node still regards it as
+              // connected.
+              // this not atomic check is safe as long as incoming connections are processed one by one, and outgoing and
+              // incoming connections are disjoint.
+              // another option of dealing with this case would be to try to replace old faulty connection with new one.
+              if (alreadyRegistered) {
+                encryptedConnection.close()
+              } else {
+                connectionsRegister
+                  .registerConnection(encryptedConnection)
+                  .flatMap(_ => connectionsQueue.offer(encryptedConnection))
+              }
+            }
+        } else {
+          encryptedConnection.close()
+        }
+      }
+      .completedL
+  }
+
+  def withCancelToken[F[_]: Concurrent, A](
+      token: Deferred[F, Unit],
+      ops: F[Option[A]]
+  ): F[Option[A]] =
+    Concurrent[F].race(token.get, ops).map {
+      case Left(()) => None
+      case Right(x) => x
+    }
+
+  private def connectionFinishHandler[F[_]: Concurrent, K, M](
+      connection: EncryptedConnection[F, K, M],
+      connectionsToAcquire: ConcurrentQueue[F, OutGoingConnectionRequest[K]],
+      connectionsRegister: ConnectionsRegister[F, K, M],
+      clusterConfig: ClusterConfig[K]
+  ): F[Unit] = {
+
+    for {
+      _ <- connection.close()
+      _ <- connectionsRegister.deregisterConnection(connection)
+      _ <-
+        if (clusterConfig.isOutGoingConnection(connection.remotePeerInfo._1)) {
```

Comment:
One difference between `close` message is that handshake is on transport layer level (tls) and `close` message is strictly application level protocol.

Anyway I am all in for some disconnect message just to help clean resources from disconnected peers. Currently we have `keep-alive` option enabled on default, but it is OS dependant and usually kicks after only 2h ( but it depends on os). We have also `close_notify` on tls level i.e each peer should send `close_notify` when disconnecting, but it is on the other hand library dependant.  From what I see here https://netty.io/4.0/api/io/netty/handler/ssl/SslHandler.html, netty uses it, but if we would change to use native implementation then we have no such guarantee.

I would do this in separate story as this pr is already big as it is (maybe along side with heart beating)

Reply:
just fyi, for some hearbeating, netty have a special handler to generate `ChanneldleEvents` if there was not write, read or both on some channel, it could be good config option addition to scalanet.

Reply:
Cool, you always find good features in Netty. Agreed it's not for this PR.